File: javascript/hashtable/hashtable.js
export default class HashTable {
    constructor(size = 10) {
        this.table = new Array(size).fill(null).map(() => []);
    }

    hash(key) {
        let total = 0;
        for(let i = 0; i < key.length; i++) {
            total += key.charCodeAt(i);
        }
        return total % this.table.length;
    }

    put(key, value) {
        const index = this.hash(key);
        const bucket = this.table[index];
        
        for(let i = 0; i < bucket.length; i++) {
            if(bucket[i][0] === key) {
                bucket[i][1] = value;
                return;
            }
        }
        bucket.push([key, value]);
    }

    get(key) {
        const index = this.hash(key);
        const bucket = this.table[index];
        
        for(const [k, v] of bucket) {
            if(k === key) {
                return v;
            }
        }
        return undefined;
    }

    remove(key) {
        const index = this.hash(key);
        const bucket = this.table[index];
        
        for(let i = 0; i < bucket.length; i++) {
            if(bucket[i][0] === key) {
                bucket.splice(i, 1);
                return true;
            }
        }
        return false;
    }

    toString() {
        return this.table.map((bucket, i) => {
            const pairs = bucket.map(([k, v]) => `(${k}:${v})`).join(' ');
            return `Bucket[${i}]: ${pairs}`;
        }).join('\n');
    }
}

File: scala/hashtable/src/main/scala/Hashtable.scala

trait HashTable[K, V]:
  def put(key: K, value: V): Unit
  def get(key: K): Option[V]
  def remove(key: K): Boolean
  def isEmpty: Boolean

case class Entry[K, V](key: K, var value: V)

// size should be bigger and prime
// using small for testing
class HashTableImpl[K, V](size: Int = 10) extends HashTable[K, V]:
  private val table: Array[List[Entry[K, V]]] = Array.fill(size)(List.empty)

  // should be better hashing, good hashing has less collision
  private def hash(key: K): Int = 
    Math.abs(key.hashCode % size)

  override def put(key: K, value: V): Unit =
    val index = hash(key)
    val bucket = table(index)
    
    bucket.find(_.key == key) match
      case Some(entry) => entry.value = value
      case None => table(index) = Entry(key, value) :: bucket

  override def get(key: K): Option[V] =
    val bucket = table(hash(key))
    bucket.find(_.key == key).map(_.value)

  override def remove(key: K): Boolean =
    val index = hash(key)
    val bucket = table(index)
    
    bucket.find(_.key == key) match
      case Some(_) =>
        table(index) = bucket.filterNot(_.key == key)
        true
      case None => false

  override def isEmpty: Boolean =
    table.forall(_.isEmpty)

  override def toString: String =
    table.zipWithIndex.map { (bucket, i) =>
      val entries = bucket.map(e => s"(${e.key}:${e.value})").mkString(" ")
      s"Bucket[$i]: $entries"
    }.mkString("\n")